src/lc3_vm/virtual_machine.rs

```diff
@@ -143,108 +158,267 @@ impl VM {
+                // The address of the value is calculated by adding the incremented PC to the
+                // sign extended offset
+                let address = pc_value.wrapping_add(offset);
+                // Read the word from the memory address
+                let word = self
+                    .read_word(address)
+                    .map_err(|err| VMError::Execute(format!("LD: {}", err)))?
+                    .ok_or(VMError::Execute(String::from("LD: read_word")))?;
+                // Store the word into the destination register
+                self.update_register(dr.into(), word)
+                    .map_err(|err| VMError::Execute(format!("LD: {}", err)))?;
+
+                self.update_flags(dr.into())
+                    .map_err(|err| VMError::Execute(format!("LD: {}", err)))?;
             }
             Opcode::ST { sr, offset } => {
-                println!("{sr}");
-                println!("{offset}");
+                // Get the word to store from the source register
+                let word = self
+                    .get_register_value(sr.into())
+                    .map_err(|err| VMError::Execute(format!("ST: {}", err)))?;
+
+                let pc_value = self
+                    .get_pc()
+                    .map_err(|err| VMError::Execute(format!("ST: {}", err)))?;
+                let offset = sign_extend_9_bits(offset);
+                // The address of the value is calculated by adding the incremented PC to the
+                // sign extended offset
+                let address = pc_value.wrapping_add(offset);
+                // Store the word into the calculated memory address
+                self.store_word(address, word)
+                    .map_err(|err| VMError::Execute(format!("ST: {}", err)))?;
             }
             Opcode::JSR { mode, offset } => {
-                println!("{mode}");
-                print!("{offset}");
+                // Save PC into R7
+                let pc_value = self
+                    .get_pc()
+                    .map_err(|err| VMError::Execute(format!("JSR: {}", err)))?;
+                self.update_register(7, pc_value)
+                    .map_err(|err| VMError::Execute(format!("JSR: {}", err)))?;
+                // Calculate offset depending on mode
+                let new_pc_value = if mode {
+                    // If the mode flag is set the PC is the sum of the incremented PC and the sign
+                    // extended offset
+                    pc_value.wrapping_add(sign_extend_11_bits(offset))
+                } else {
+                    // If the mode flag is not set the pc is the base register, we shift the value 6 times
+                    // to the right because the base address is stored in the 3 most significant bits of the offset
+                    self.get_register_value(offset >> 6)
+                        .map_err(|err| VMError::Execute(format!("JSR: {}", err)))?
+                };
```

Comment:
I think it would be cleaner if this logic can be determined in the decoding phase and take advantage of the type system, you can divide between `JSR` and `JSRR` instructions both with different arguments.

Reply:
For now I won't change it but I'll have it in mind for the future.